File: tests/test_fs_id_map.c
/*
 * test_fs_id_map.c — standalone unit test for the fs_list.h backend-id surface
 * (brix_fs_id_from_name / brix_fs_id_name).
 *
 * Build+run (no nginx tree needed — sd_fs_id.c is ngx-free):
 *   gcc -I src -o /tmp/test_fs_id_map tests/test_fs_id_map.c src/fs/backend/sd_fs_id.c && /tmp/test_fs_id_map
 */
#include <stdio.h>
#include <string.h>

#include "core/types/fs_list.h"

static int failures = 0;

static void
check(int cond, const char *what)
{
    if (cond) {
        printf("  ok   %s\n", what);
    } else {
        printf("  FAIL %s\n", what);
        failures++;
    }
}

int
main(void)
{
    int  id;
    char label[64];

    /* Roundtrip every census row: name -> id -> name. */
    for (id = 0; id < BRIX_FS_ID_COUNT; id++) {
        const char *name = brix_fs_id_name(id);

        snprintf(label, sizeof(label), "row %d roundtrip (%s)", id, name);
        check(name != NULL && name[0] != '\0'
              && brix_fs_id_from_name(name) == id, label);
    }

    /* Known anchors present in every build. */
    check(brix_fs_id_from_name("posix") >= 0, "posix registered");
    check(brix_fs_id_from_name("cache") >= 0, "cache decorator registered");
    check(brix_fs_id_from_name("xroot") >= 0, "xroot origin registered");

    /* Negatives: unknown / NULL / out-of-range id. */
    check(brix_fs_id_from_name("nosuchfs") == -1, "unknown name -> -1");
    check(brix_fs_id_from_name(NULL) == -1, "NULL name -> -1");
    check(strcmp(brix_fs_id_name(-1), "?") == 0, "id -1 -> \"?\"");
    check(strcmp(brix_fs_id_name(BRIX_FS_ID_COUNT), "?") == 0,
          "id COUNT -> \"?\"");

    printf(failures == 0 ? "ALL PASS\n" : "%d FAILURES\n", failures);
    return failures == 0 ? 0 : 1;
}

File: tests/test_fs_id_map.py
"""Surface for tests/test_fs_id_map.c — the fs_list.h backend-census unit test.

The census (core/types/fs_list.h) is THE single declaration of every storage
driver: the name→driver registry, the per-backend SHM metric arrays, and the
exporter labels all generate from it. This test roundtrips every row
(name → id → name), pins the always-present anchors (posix/cache/xroot/frm and
the wave-34 `mirage` synthetic backend), and enforces the append-only rule
(posix stays row 0 — a renumbering would silently re-key the SHM byte
counters across a reload).

The C test was previously ORPHANED (no surface ran it); it builds standalone —
`sd_fs_id.c` is ngx-free — exactly per its own header comment: no nginx tree,
no objs, just cc.
"""

import os
import subprocess

import pytest

_REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def test_fs_id_census_roundtrip(tmp_path):
    cc = os.environ.get("CC", "cc")
    exe = tmp_path / "test_fs_id_map"
    build = subprocess.run(
        [cc, "-Wall", "-Werror", "-I", os.path.join(_REPO, "src"),
         "-o", str(exe),
         os.path.join(_REPO, "tests", "test_fs_id_map.c"),
         os.path.join(_REPO, "src", "fs", "backend", "sd_fs_id.c")],
        capture_output=True, text=True, timeout=60)
    if build.returncode != 0 and "not found" in build.stderr.lower():
        pytest.skip(f"no C compiler: {build.stderr.splitlines()[:1]}")
    assert build.returncode == 0, f"census unit test failed to build:\n{build.stderr}"

    run = subprocess.run([str(exe)], capture_output=True, text=True, timeout=30)
    assert run.returncode == 0, f"census checks failed:\n{run.stdout}{run.stderr}"
    assert "ALL PASS" in run.stdout, run.stdout
